### PR TITLE
KAFKA-16790: Update RemoteLogManager configuration in broker server

### PR DESCRIPTION
**Pull request by muralibasani**
Details of the issue is defined in https://issues.apache.org/jira/browse/KAFKA-16790

In BrokerServer.scala, brokerMetadataPublishers are configured and when there are metadata updates remoteLogManager is not configured by then. 
Ex  : remoteLogManager.foreach(rlm => rlm.onLeadershipChange(partitionsBecomeLeader.asJava, partitionsBecomeFollower.asJava, topicIds)) in ReplicaManager is invoked after publishers are instantiated, and here rlm has relevant managers configured.

This change makes sure rlm is configured before the brokerMetadataPublishers initialization.

*Summary of testing strategy (including rationale)
for the feature or bug fix. Unit and/or integration
tests are expected for any behaviour change and
system tests should be considered for larger changes.*

### Committer Checklist (excluded from commit message)
- [X] Verify design and implementation 
- [X] Verify test coverage and CI build status
- [X] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by muralibasani** [comment]
@clolov would you like to take a look ? I have taken the first approach of configuring rlm before initializing publishers, instead of skipping if they don't exist. There can be multiple places where rlm configuration is required. So thought this is better.

Haven't added any explicit tests here, but do you suggest any ?

**Comment by clolov** [comment]
This approach makes sense to me. How easy would it be to code up the test which I did manually in the JIRA ticket? In other words, without this change I would expect the test to fail with a NullPointerException (or an exception of your choice)

**Comment by muralibasani** [comment]
@clolov thanks for the tip.
Would it be better to write the test in KRaftClusterTest.scala or RemoteLogManagerTest or in a different file ? 
As it's more of an integration test, where broker starts up, and if we mock it without rlm initialization, it would fail with NullPointer proobably.

**Comment by muralibasani** [comment]
And I tried testing with export INCLUDE_TEST_JARS=true, but still notice class not found err. Am I missing any ?

**Comment by muralibasani** [comment]
@clolov 
In ReplicaManager.scala, 
if REMOTE_LOG_STORAGE_SYSTEM_ENABLE_PROP is enabled and if remoteLogManager is None
do we want to throw any exception before the below line
remoteLogManager.foreach(rlm => rlm.onLeadershipChange(l.... 

then we can add a test easily

**Comment by muralibasani** [comment]
@clolov I tried to add 
        .setConfigProp(RemoteLogManagerConfig.REMOTE_LOG_STORAGE_SYSTEM_ENABLE_PROP, true.toString)
        .setConfigProp(RemoteLogManagerConfig.REMOTE_LOG_METADATA_MANAGER_CLASS_NAME_PROP,
          "org.apache.kafka.server.log.remote.storage.NoOpRemoteLogMetadataManager")
        .setConfigProp(RemoteLogManagerConfig.REMOTE_STORAGE_MANAGER_CLASS_NAME_PROP,
          "org.apache.kafka.server.log.remote.storage.NoOpRemoteStorageManager")
to this test here
 https://github.com/apache/kafka/blob/trunk/core/src/test/scala/integration/kafka/server/KRaftClusterTest.scala#L163

What  I noticed is remoteLogManager at https://github.com/apache/kafka/blob/trunk/core/src/main/scala/kafka/server/ReplicaManager.scala#L2737 is not null and it is configured with managers, as broker is fully started even with old code. This is confusing now.

**Comment by nikramakrishnan** [comment]
Hi @muralibasani, thanks for the PR!

As I understand it, the issue here is that `rlm.startup()` (which internally calls `remoteLogMetadataManager.configure`) is not called first when there is metadata to catch up on during broker startup. This means that RemoteLogManager may be non-null, but `applyDelta` ends up calling `rlm.onLeadershipChange` (which calls `remoteLogMetadataManager.onPartitionLeadershipChanges`) before `configure` is invoked on `remoteLogMetadataManager`.

A possible way to test this could be to mock `configure` and `onPartitionLeadershipChanges` in `RemoteLogMetadataManager`, then call `applyDelta` with some leadership changes, and assert that configure was called on the mock before `onPartitionLeadershipChanges`.

**Comment by muralibasani** [comment]
@nikramakrishnan 
I think calling applyDelta on replicaManager wouldn't invoke configure on rlmm. Probably we should start mocking from BrokerServer.startup method, and create a topic, which invoke both configure and applyDelta
wdyt ?

**Comment by nikramakrishnan** [comment]
> @nikramakrishnan I think calling applyDelta on replicaManager wouldn't invoke configure on rlmm. Probably we should start mocking from BrokerServer.startup method, and create a topic, which invoke both configure and applyDelta wdyt ?

Agree to this. Maybe we could do something similar to what is being done in [DeleteTopicTest.createTestTopicAndCluster](https://github.com/apache/kafka/blob/trunk/core/src/test/scala/unit/kafka/admin/DeleteTopicTest.scala#L451)?

**Comment by muralibasani** [comment]
@nikramakrishnan somehow am not feeling confident to write a test based on the order of execution. Instead we should throw an exception (or stateChangeLogger.error) like mentioned below here, and write a unit test just for this class. With this approach, whenever there are metadata changes, and remote log storage is enabled. wdyt ?

> @clolov In ReplicaManager.scala, if REMOTE_LOG_STORAGE_SYSTEM_ENABLE_PROP is enabled and if remoteLogManager is None do we want to throw any exception before the below line remoteLogManager.foreach(rlm => rlm.onLeadershipChange(l....
> 
> then we can add a test easily

**Comment by nikramakrishnan** [comment]
> Instead we should throw an exception (or stateChangeLogger.error) like mentioned below here, and write a unit test just for this class

Do you mean we mock `onPartitionLeadershipChanges` to throw an exception if it is invoked without `configure` being called? We can take that approach as well.

**Comment by muralibasani** [comment]
@nikramakrishnan I have pushed a few changes. Please review.
We explicitly throw a KafkaException now if rlm is not configured and remote storage is enabled

File: core/src/test/scala/unit/kafka/server/ReplicaManagerTest.scala
Comment (showuon) [inline core/src/test/scala/unit/kafka/server/ReplicaManagerTest.scala]: nit: indent is wrong.

File: core/src/test/scala/integration/kafka/server/KRaftClusterTest.scala
Comment (showuon) [inline core/src/test/scala/integration/kafka/server/KRaftClusterTest.scala]: nit: 1 broker + 1 controller should be enough for this test scenario, right?

**Review by showuon — COMMENTED** [review]
@muralibasani , thanks for the PR. Left some comments.

File: core/src/test/scala/unit/kafka/server/ReplicaManagerTest.scala
Comment (muralibasani) [reply core/src/test/scala/unit/kafka/server/ReplicaManagerTest.scala]: Ah yes. fixed.

File: core/src/test/scala/integration/kafka/server/KRaftClusterTest.scala
Comment (muralibasani) [reply core/src/test/scala/integration/kafka/server/KRaftClusterTest.scala]: agree, we don't need. fixed

**Comment by muralibasani** [comment]
@showuon can we still make this for 3.8.0, I think the deadline is today.

**Comment by showuon** [comment]
> @showuon can we still make this for 3.8.0, I think the deadline is today.

The feature freeze date is related to KIP, not bug fix. So no worries.

File: core/src/main/java/kafka/log/remote/RemoteLogManager.java
Comment (showuon) [inline core/src/main/java/kafka/log/remote/RemoteLogManager.java]: nit: Should we move this validation on the top of `onLeadershipChange` method?

**Review by showuon — COMMENTED** [review]
LGTM! Just one minor comment.

File: core/src/main/java/kafka/log/remote/RemoteLogManager.java
Comment (muralibasani) [reply core/src/main/java/kafka/log/remote/RemoteLogManager.java]: Agree, we can throw before deriving other vars. Moved.

**Review by showuon — APPROVED** [review]
LGTM! Will wait for CI build completes.

**Comment by muralibasani** [comment]
> LGTM! Will wait for CI build completes.

looks like ci is completed with warnings

**Comment by nikramakrishnan** [comment]
Thanks for the changes @muralibasani, and sorry for the delay in getting back to this. These changes LGTM!

There seem to be failing tests related to this change though:

`testRemoteLogTaskUpdateRemoteLogSegmentMetadataAfterLogDirChanged` fails with:

```
org.apache.kafka.common.KafkaException: RemoteLogManager is not configured when remote storage system is enabled
``` 

Can you please check this?

**Comment by muralibasani** [comment]
> Thanks for the changes @muralibasani, and sorry for the delay in getting back to this. These changes LGTM!
> 
> There seem to be failing tests related to this change though:
> 
> `testRemoteLogTaskUpdateRemoteLogSegmentMetadataAfterLogDirChanged` fails with:
> 
> ```
> org.apache.kafka.common.KafkaException: RemoteLogManager is not configured when remote storage system is enabled
> ```
> 
> Can you please check this?

thanks. there seems to be a new test added yesterday. Updated it.

**Review by nikramakrishnan — APPROVED** [review]
Thanks for addressing the comments, LGTM assuming the tests pass!

**Comment by muralibasani** [comment]
@nikramakrishnan @showuon there are some test failures but not sure if related to ours ?

**Comment by showuon** [comment]
Failed tests are unrelated.

**Review by showuon — APPROVED** [review]

**Comment by clolov** [comment]
Thanks for persevering with this @muralibasani!

**Pull request merged**
